### PR TITLE
Added tutorial for Tarjan's Algorithm

### PR DESCRIPTION
**Pull request by BlackPanther112358**
## Change Summary

Added Tutorial for Tarjan's algorithm

## Checklist

**If you haven't fulfilled the below requirements or even delete the entire checklist, your PR won't be reviewed and will be closed without notice.**

### General 

- [x] This Pull Request is all my own work. (You'll be blacklisted if you are caught for plagiarism.)
- [x] I've read [CONTRIBUTING.md](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md)
- [x] I've applied LaTex for all variables, formulas and time / space complexity instead of using backticks
- [x] I've started the app locally and verified all the content and all links (if applicable) are accessible correctly
- [x] I've included Complexity Analysis (Time Complexity & Space Complexity).
- [x] I've written my explanation well and it is easy to understand for beginners

### Tutorial

- [x] I've read and followed the [Tutorial Template](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md#tutorial-template)
- [x] I've explained my topic well with 2 - 3 LC problems and no external problems are used. 
- [x] I've provided the full working solutions to the problems used in this tutorial.
- [x] I've provided suggested problems at the end with the given format. See [here](https://raw.githubusercontent.com/wingkwong/leetcode-the-hard-way/main/tutorials/math/number-theory/binary-exponentiation.md) as an example. If the target solution is not available, leave `solutionLink` blank.
- [x] I've given credits / references if I use external resources. (For an image, give credit under it. Otherwise, add a new section called References at the end (after Suggested Problems).)

### Solutions

- [x] I've read and followed the [Solution Template](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md#solution-template)
- [x] I've formatted my code well with [K&R Coding style](https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr)
- [x] I've confirmed that comments are put above each line rather than writing on the same line.
- [ ] I've included a meaningful approach name for my solution. e.g. `## Approach 1: Two Pointers`.

### CODE REVIEW

**Comment by BlackPanther112358** [comment]
Hi, this is for Issue number #67, apologies I forgot to add to title

File: tutorials/graph-theory/tarjans-algorithm.md [hunk 1]
```diff
@@ -1,11 +1,346 @@
...
+
+![image](https://user-images.githubusercontent.com/83649829/198752836-3c43a6f4-e1e2-445e-b679-0df4cb8ff3e0.png)$\\$
+[Source](https://en.wikipedia.org/wiki/Strongly_connected_component)
+
+
+In this tutorial we will discuss the Tarjan's Algorithm to find SCC.
+
+## Implementation
+
+The algorithm works by using a single [DFS](../../tutorials/graph-theory/depth-first-search) in the graph. To understand its working first let's define a few standard terms :-
```
Comment (wingkwong) [inline tutorials/graph-theory/tarjans-algorithm.md:27]: `:-` -> `:`

File: tutorials/graph-theory/tarjans-algorithm.md [hunk 2]
```diff
@@ -1,11 +1,346 @@
...
+
+<TutorialAuthors names="@BlackPanther112358"/>
+
+## Overview
+
+Tarjan's Algorithm is popular algorithm for finding the Strongly Connected Components (SCC) of a directed graph. In the below graph, the nodes in a blue envelope constitute a single connected component as any node $u$ as a path to another node $v$ and vice versa. Note that while we can reach the node $f$ from the node $e$, the opposite is not true. Hence, $f$ and $e$ are not part of the same component. Thus, the following graph has 3 strongly connected components as highlighted. A single strongly connected component can be formally described as maximal set of vertices such that for any 2 vertices belonging to the set, say $u$ and $v$, there exists a path from $u$ to $v$ and vice versa.
+
+
+![image](https://user-images.githubusercontent.com/83649829/198752836-3c43a6f4-e1e2-445e-b679-0df4cb8ff3e0.png)$\\$
+[Source](https://en.wikipedia.org/wiki/Strongly_connected_component)
```
Comment (wingkwong) [inline tutorials/graph-theory/tarjans-algorithm.md:20]: Source: Strongly connected component from Wikipedia

File: tutorials/graph-theory/tarjans-algorithm.md [hunk 3]
```diff
@@ -1,11 +1,346 @@
...
+
+$low(u) =$ smallest time reachable of a node reachable from the DFS subtree of node u
+
+We also need to ensure that we don't mix 2 different SCC's due to a cross-edge between them. To counter this issue, we will use a stack to store the nodes which have not been assigned to an SCC and have been visited so far.
+
+Thus, whenever we find an SCC, we will pop the corresponding nodes from the stack.
+
+We will call the first node we discovered of an SCC the route node for sake of simplicity. Thus, we will identify the SCC by its root node. To check when we have reached a root node, we just check if $low(u)$ and $tin(u)$ are equal.
+
+The pseudo-code for the algorithm can be found [here](https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm#The_algorithm_in_pseudocode)
```
Comment (wingkwong) [inline tutorials/graph-theory/tarjans-algorithm.md:39]: add a full stop

File: tutorials/graph-theory/tarjans-algorithm.md [hunk 4]
```diff
@@ -1,11 +1,346 @@
...
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+struct tarjans_algo{
+    // Recursive function to do the DFS search on the graph
+    void dfs(int u, int &timer, vector<int> &tin, vector<int> &low_link, vector<bool> &on_stack, stack<int> &stk, vector<vector<int>> &graph, vector<vector<int>> &res){
```
Comment (wingkwong) [inline tutorials/graph-theory/tarjans-algorithm.md:54]: this line is too long. ideally we don't let readers to scroll

File: tutorials/graph-theory/tarjans-algorithm.md [hunk 5]
```diff
@@ -1,11 +1,346 @@
...
+The implementation of above can be as follows (along with above graph as example) :
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+struct tarjans_algo{
```
Comment (wingkwong) [inline tutorials/graph-theory/tarjans-algorithm.md:52]: prefer a better name

File: tutorials/graph-theory/tarjans-algorithm.md [hunk 6]
```diff
@@ -1,11 +1,346 @@
...
+    vector<vector<int>> res = t.tarjans(graph);
+
+    // Output the final result
+    cout << "The Strongly Connected Components for the graph are:-" << endl;
+    for (vector<int> i : res) {
+        for (int j : i) {
+            cout << (char)('a' + j) << " ";
+        }
+        cout << endl;
+    }
```
Comment (wingkwong) [inline tutorials/graph-theory/tarjans-algorithm.md:154]: show the result also. supposedly readers won't copy your code and run locally

File: tutorials/graph-theory/tarjans-algorithm.md [hunk 7]
```diff
@@ -1,11 +1,346 @@
...
+    graph[6].push_back(5);
+    graph[7].push_back(3);
+    graph[7].push_back(6);
+
+    // Using the tarjan's algo
+    tarjans_algo t = tarjans_algo();
+    vector<vector<int>> res = t.tarjans(graph);
+
+    // Output the final result
+    cout << "The Strongly Connected Components for the graph are:-" << endl;
```
Comment (wingkwong) [inline tutorials/graph-theory/tarjans-algorithm.md:148]: `:-` -> `:`

File: tutorials/graph-theory/tarjans-algorithm.md [hunk 8]
```diff
@@ -1,11 +1,346 @@
...
+[Source](https://en.wikipedia.org/wiki/Strongly_connected_component)
+
+
+In this tutorial we will discuss the Tarjan's Algorithm to find SCC.
+
+## Implementation
+
+The algorithm works by using a single [DFS](../../tutorials/graph-theory/depth-first-search) in the graph. To understand its working first let's define a few standard terms :-
+
+$tin(u) =$ time at which node $u$ was reached for the first time or the in time of DFS for the node.
```
Comment (wingkwong) [inline tutorials/graph-theory/tarjans-algorithm.md:29]: prefer a better name. $tin$ is not intuitive

File: tutorials/graph-theory/tarjans-algorithm.md [hunk 9]
```diff
@@ -1,11 +1,346 @@
...
+
+        tin.resize(n, -1);
+        low_link.resize(n, -1);
+        graph.resize(n);
+
+        // Make an adjacency list from the input
+        for (auto i : connections) {
+            graph[i[0]].push_back(i[1]);
+            graph[i[1]].push_back(i[0]);
+        }
```
Comment (wingkwong) [inline tutorials/graph-theory/tarjans-algorithm.md:222]: `i` -> `connection`

**Review by wingkwong — CHANGES_REQUESTED** [review]
Overall it looks good.

1. comment starts with lower case
2. follow [K&R Coding style](https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr)

File: tutorials/graph-theory/tarjans-algorithm.md (hunk 4 above)
Comment (BlackPanther112358) [reply tutorials/graph-theory/tarjans-algorithm.md:54]: I need all of these datastructures to run the algorithm, the other option is to make these global, which I am doing for now. If you have something else in mind, do let me know

Comment (wingkwong) [reply tutorials/graph-theory/tarjans-algorithm.md]: Either breaking them into multiple lines or making some of them global

Comment (BlackPanther112358) [reply tutorials/graph-theory/tarjans-algorithm.md]: I have made all of them except adjacency list global, do have a look once you find time

**Review by wingkwong — APPROVED** [review]

**Pull request merged**